addons/project_purchase/models/project.py

```diff
@@ -139,4 +143,44 @@ def _get_profitability_items(self, with_action=True):
                 costs['data'].append(purchase_order_costs)
                 costs['total']['billed'] += amount_invoiced
                 costs['total']['to_bill'] += amount_to_invoice
+            # calculate the cost of bills without a purchase order
+            query = self.env['account.move.line'].sudo()._search([
+                ('analytic_account_id', 'in', self.analytic_account_id.ids),
+                ('move_id.move_type', 'in', ['in_invoice', 'in_refund']),
+                ('parent_state', 'in', ['draft', 'posted']),
+                ('price_subtotal', '>', 0),
+                ('id', 'not in', purchase_order_line_invoice_line_ids)
+            ])
+            # account_move_line__move_id is the alias of the joined table account_move in the query
+            # we can use it, because of the "move_id.move_type" clause in the domain of the query, which generates the join
+            # this is faster than a search_read followed by a browse on the move_id to retrieve the move_type of each account.move.line
+            query_string, query_param = query.select('price_subtotal', 'parent_state', 'account_move_line__move_id.move_type')
+            self._cr.execute(query_string, query_param)
+            bills_move_line_read = [moves for moves in self._cr.dictfetchall()]
```

Comment:
the result of `self._cr.dictfetchall()` is not already a dict? 🤔

Reply:
Yeah you are right, I did some copy pasta from v16 without paying attention that it already returns a list of dict of the results. (I assumed it was an iterator that returned dicts, therefor the comprehension was necessary, my bad)

https://github.com/odoo/odoo/blob/8947fd1251874454bdfa6508fb32d2b44afe1f09/addons/project_purchase/models/project.py#L131